File: src/middle/patterns.jl
# Partial Pattern Matching

# Match types (partial values inferred by the interpreter) against user-
# specified patterns at compile time.
# Returns either a dictionary (if the match is certain to succeed), `nothing`
# (if the match is certain to fail) or `missing` (we can't handle the match
# statically).
#
# The returned dictionary contains a mapping from binding names to (type, path)
# pairs, where `path` tells the compiler where to find the matched value in the
# original data. `path` is a list of indexes; the last index might be a
# UnitRange, representing a splat.
#
# Ideally we'd think of dispatchers as a simple list of `if` clauses: for each
# method, check the arg list against the signature via `match(sig, args)`, and
# if the match succeeds call the method. Then we do type inference and remove
# redundant checks using the usual, generic optimisations.
#
# However, `match` is not available until it's defined in the stdlib, for which
# we need function calls (and thus dispatchers, and thus `match`) to work. To
# break the cycle we take a shortcut in some simple cases; if an input obviously
# matches the signature, we'll generate code for the dispatcher that behaves
# like `match`.

function trym(x)
  @q begin
    local result = $(esc(x))
    (result === nothing || result === missing) && return result
    result
  end
end

@eval macro $:try(x)
  trym(x)
end

_assoc(::Missing, _) = missing

function _assoc(as, (name, (val, path)))
  haskey(as, name) || return merge(as, Dict(name => (val, path)))
  val′ = as[name][1]
  if isvalue(rlist(val′, val))
    return val′ == val ? as : nothing
  else
    return missing # could reject more cases here
  end
end

function _merge(as, bs)
  (ismissing(as) || ismissing(bs)) && return missing
  for b in bs
    as = @try _assoc(as, b)
  end
  return as
end

function partial_match(mod, pat::Hole, val, path)
  return Dict()
end

function partial_match(mod, pat::Literal, val, path)
  if isvalue(val)
    return pat.value == val ? Dict() : nothing
  else
    return missing # could reject more cases here
  end
end

function partial_match(mod, pat::Bind, val, path)
  bs = @try partial_match(mod, pat.pattern, val, path)
  return _assoc(bs, pat.name => (val, path))
end

function partial_match(mod, pat::Trait, val, path)
  r = trivial_isa(mod, val, pat.pattern)
  r === true ? Dict() : r === false ? nothing : missing
end

function partial_match(mod, pat::And, val, path)
  bs = Dict()
  for p in pat.patterns
    bs = @try _merge(bs, @try partial_match(mod, p, val, path))
  end
  return bs
end

function partial_match(mod, pat::Or, val, path)
  for p in pat.patterns
    m = partial_match(mod, p, val, path)
    isnothing(m) && continue
    ismissing(m) && return missing
    return m
  end
  return
end

function partial_match(mod, pat::Constructor, val, path)
  pat = rvpattern(pat)
  pat = mod[(tag"common.constructorPattern", rlist(parts(pat)...))]
  isvalue(pat) || return missing
  return partial_match(mod, jlpattern(part(pat, 1)), val, path)
end

ishole(x::Hole) = true
ishole(x::Bind) = ishole(x.pattern)
ishole(x) = false

isslurp(x) = x isa Repeat && ishole(x.pattern)

function partial_match(mod, pat::Pack, val::SimpleType, path)
  bs = Dict()
  i = 0
  while true
    i <= nparts(val) || break
    i <= nparts(pat) || return
    if pat[i] == Repeat(hole)
      break
    elseif isslurp(pat[i])
      bs = @try _assoc(bs, pat[i].pattern.name => (rlist(allparts(val)[i+1:end]...), [path..., i:nparts(val)]))
      return bs
    elseif pat[i] isa Repeat
      return missing
    else
      # continue on `missing`, since we might narrow to `nothing` later
      b = partial_match(mod, part(pat, i), part(val, i), [path..., i])
      b === nothing && return
      bs = _merge(bs, b)
    end
    i += 1
  end
  if nparts(pat) == i && isslurp(pat[i])
    if pat[i].pattern != hole
      bs = @try _assoc(bs, pat[i].pattern.name => (rlist(), [path..., i:0]))
    end
  elseif nparts(pat) > nparts(val)
    return
  end
  return bs
end

# TODO match results don't have to be identical, if
# bindings and paths are right we can merge types.
function partial_match_union(mod, pat, val::Onion, path)
  ms = map(x -> partial_match(mod, pat, x, path), val.types)
  any(x -> x === missing, ms) && return missing
  all(==(first(ms)), ms) && return first(ms)
  return missing
end

function partial_match(mod, pat::Pack, val::Onion, path)
  partial_match_union(mod, pat, val, path)
end

function partial_match(mod, pat::Or, val::Onion, path)
  partial_match_union(mod, pat, val, path)
end

function partial_match(mod, pat::Pack, val::Recursive, path)
  partial_match(mod, pat, unroll(val), path)
end

function partial_match(mod, pat::Or, val::Recursive, path)
  partial_match(mod, pat, unroll(val), path)
end

function partial_match(mod, pat::Pack, val::VPack, path)
  bs = @try partial_match(mod, tag(pat), tag(val), [path..., 0])
  isempty(bs) || return missing
  (nparts(pat) == 1 && part(pat, 1) isa Repeat) || return missing
  pat = part(pat, 1).pattern
  b, r = pat isa Bind ? (pat.name, pat.pattern) : (nothing, pat)
  bs′ = partial_match(mod, r, val.parts, path)
  isnothing(bs′) && return
  isempty(bs′) || return missing
  return b == nothing ? bs : _assoc(bs, b => (val, path))
end

partial_match(mod, pat, val) = partial_match(mod, pat, val, [])

# TODO assumes the value is unchanged by the match
function trivial_isa(int, val, T)
  r = int[(tag"common.matchTrait", rlist(T, val))]
  isnothing(r) && return missing
  T = tag(part(r, 1))
  T == tag"common.Some" ? true :
  T == tag"common.Nil" ? false :
  missing
end

# Filtered methods

function MatchMethods(defs, interp)
  EagerCache() do _, (f, Ts)
    result = []
    for meth in reverse(defs[f])
      m = partial_match(interp, meth.sig.pattern, Ts)
      if isnothing(m)
        continue
      else
        push!(result, (meth, m))
        ismissing(m) || break
      end
    end
    return result
  end
end

# Generate dispatchers

dispatch_arms(T) = [T]
dispatch_arms(T::Onion) = T.types
dispatch_arms(T::Recursive) = dispatch_arms(unroll(T))

function dispatch_arms(T::Pack)
  result = [[]]
  for part in dispatch_arms.(T.parts)
    result = [[prefix..., x] for prefix in result for x in part]
  end
  return [pack(x...) for x in result]
end

function indexer!(ir::IR, T, arg, path)
  isempty(path) && return arg
  (p, rest...) = path
  if p isa AbstractVector
    ps = [push!(ir, stmt(xpart(arg, i), type = part(T, i))) for i in p]
    arg = push!(ir, stmt(xlist(ps...), type = rlist(exprtype(ir, ps)...)))
  else
    T = part(T, p)
    arg = push!(ir, stmt(xpart(arg, p), type = T))
  end
  arg = indexer!(ir, T, arg, rest)
end

function icall!(inf, ir, sig, f, args...)
  if !(f isa RMethod)
    args = [push!(ir, stmt(xlist(args...), type = rlist(exprtype(ir, args)...)))]
  end
  ex = xcall(f, args...)
  T = infercall!(inf, sig, ir, ex)
  push!(ir, stmt(ex, type = T))
end

function dispatcher(inf, func::Tag, Ts)
  ir = IR(meta = FuncInfo(func, trampoline = true))
  args = argument!(ir, type = Ts)
  ret = ⊥
  arms = dispatch_arms(Ts)
  call!(f, args...) = icall!(inf, ir, (func, Ts), f, args...)
  for (meth, m) in inf.meths[(func, Ts)]
    if ismissing(m)
      pat = rvpattern(meth.sig.pattern)
      arms = filter(T -> issubset(rlist(nil), infercall!(inf, (func, Ts), tag"common.match", rlist(T, pat))), arms)
      m = call!(tag"common.match", args, pat)
      m = call!(part_method, m, 1)
      cond = push!(ir, stmt(xcall(isnil_method, m), type = partial_isnil(exprtype(ir, m))))
      branch!(ir, length(blocks(ir))+2, when = cond)
      branch!(ir, length(blocks(ir))+1)
      block!(ir)
      m = call!(notnil_method, m)
      as = []
      for arg in meth.sig.args
        push!(as, call!(part_method, call!(tag"common.getkey", m, Tag(arg)), 1))
      end
      result = call!(meth, as...)
      isempty(meth.sig.swap) && (result = push!(ir, stmt(xlist(result), type = rlist(exprtype(ir, result)))))
      return!(ir, result)
      ret = union(ret, exprtype(ir, result))
      block!(ir)
    else # certain to match
      result = call!(meth, [indexer!(ir, Ts, args, m[x][2]) for x in meth.sig.args]...)
      isempty(meth.sig.swap) && (result = push!(ir, stmt(xlist(result), type = rlist(exprtype(ir, result)))))
      return!(ir, result)
      ret = union(ret, exprtype(ir, result))
      return ir, ret
    end
    if isempty(arms)
      unreachable!(ir)
      return ir, ret
    end
  end
  if func == tag"common.abort" && issubset(Ts, rlist(String))
    error("Compiler fault: couldn't guarantee abort method matches")
  end
  if options().jspanic
    call!(tag"common.abort", "No matching method: $func: $Ts")
  end
  unreachable!(ir)
  return ir, ret
end

File: src/middle/patterns.ts
// Partial Pattern Matching

// Match types (partial values inferred by the interpreter) against user-
// specified patterns at compile time.
// Returns either a dictionary (if the match is certain to succeed), `nothing`
// (if the match is certain to fail) or `missing` (we can't handle the match
// statically).
//
// The returned dictionary contains a mapping from binding names to (type, path)
// pairs, where `path` tells the compiler where to find the matched value in the
// original data. `path` is a list of indexes; the last index might be a
// UnitRange, representing a splat.
//
// Ideally we'd think of dispatchers as a simple list of `if` clauses: for each
// method, check the arg list against the signature via `match(sig, args)`, and
// if the match succeeds call the method. Then we do type inference and remove
// redundant checks using the usual, generic optimisations.
//
// However, `match` is not available until it's defined in the stdlib, for which
// we need function calls (and thus dispatchers, and thus `match`) to work. To
// break the cycle we take a shortcut in some simple cases; if an input obviously
// matches the signature, we'll generate code for the dispatcher that behaves
// like `match`.

import * as types from '../frontend/types'
import * as ir from '../utils/ir'
import { MIR, IRValue, Method, Definitions } from '../frontend/modules'
import { Def } from '../dwarf'
import { xlist, xpart, xcall } from '../frontend/lower'
import { Pattern, patternType, pattern } from '../frontend/patterns'
import { Inference, Sig, inferexpr, infercall, issubset, maybe_union } from './abstract'
import { some } from '../utils/map'
import { options } from '../utils/options'
import { part_method, isnil_method, notnil_method, partial_isnil, string } from './primitives'
import { EagerCache } from '../utils/cache'
import isEqual from 'lodash/isEqual'

export { MatchMethods, partial_match, indexer, icall, dispatch_arms, dispatcher }

type Path = (number | { start: number, end: number })[]

type Match = Map<string, [types.Type, Path]>
type MatchResult = Match | null | undefined // null is known failure, undefined is unknown

interface Interpreter {
  get(func: types.Tag, args: types.Type[]): types.Type | undefined
}

function _assoc(as: Match, name: string, [val, path]: [types.Type, Path]): MatchResult {
  if (as === null || as === undefined) return as
  if (!as.has(name)) {
    const result = new Map<string, [types.Type, Path]>(as)
    result.set(name, [val, path])
    return result
  }
  const val_ = as.get(name)![0]
  if (types.isValue(types.list(val_, val))) {
    return isEqual(val_, val) ? as : null
  } else {
    return undefined // could reject more cases here
  }
}

function _merge(as: Match | undefined, bs: Match | undefined): MatchResult {
  if (as === undefined || bs === undefined) return undefined
  let m: MatchResult = as
  for (const [k, v] of bs) {
    m = _assoc(m, k, v)
    if (m === null || m === undefined) return m
  }
  return m
}

function ishole(x: Pattern): boolean {
  if (x.kind === 'hole') return true
  if (x.kind === 'bind') return ishole(x.pattern)
  return false
}

function isslurp(x: Pattern): x is Pattern & { kind: 'repeat' } {
  return x.kind === 'repeat' && ishole(x.pattern)
}

function slurpName(x: Pattern): string | undefined {
  if (x.kind === 'repeat' && x.pattern.kind === 'bind') return x.pattern.name
}

// TODO match results don't have to be identical, if
// bindings and paths are right we can merge types.
function partial_match_union(mod: Interpreter, pat: Pattern, val: types.Type & { kind: 'union' }, path: Path): MatchResult {
  const ms = val.options.map(x => partial_match(mod, pat, x, path))
  if (ms.some(x => x === undefined)) return undefined
  if (ms.every((m, i) => isEqual(m, ms[0]))) return ms[0]
  return undefined
}

function partial_match_pack(mod: Interpreter, pat: Pattern & { kind: 'pack' }, val: types.Type, path: Path): MatchResult {
  let bs: Match | undefined = new Map()
  let i = 0
  while (true) {
    if (i > types.nparts(val)) break
    if (i > pat.parts.length - 1) return null
    if (isslurp(pat.parts[i])) {
      const name = slurpName(pat.parts[i])
      if (!name || bs === undefined) return bs
      const remaining = types.allparts(val).slice(i)
      const range = { start: i, end: types.nparts(val) }
      // TODO slurps may not be at the end
      return _assoc(bs, name, [types.list(...remaining), [...path, range]])
    }
    if (pat.parts[i].kind === 'repeat') return undefined
    const b = partial_match(mod, pat.parts[i], types.part(val, i), [...path, i])
    // continue on `missing`, since we might narrow to `nothing` later
    if (b === null) return null
    const cs = _merge(bs, b)
    if (cs === null) return null
    bs = cs
    i += 1
  }
  if (pat.parts.length - 1 === i && isslurp(pat.parts[i])) {
    const name = slurpName(pat.parts[i])
    if (!name || bs === undefined) return bs
    const range = { start: i, end: 0 }
    return _assoc(bs, name, [types.list(), [...path, range]])
  } else if (pat.parts.length - 1 > types.nparts(val))
    return null
  return bs
}

function partial_match_vpack(mod: Interpreter, pat: Pattern & { kind: 'pack' }, val: types.Type & { kind: 'vpack' }, path: Path): MatchResult {
  const bs = partial_match(mod, pat.parts[0], types.tagOf(val), [...path, 0])
  if (bs === null || bs === undefined) return bs
  if (bs.size > 0) return undefined
  if (pat.parts.length !== 2 || pat.parts[1].kind !== 'repeat') return undefined
  const innerPat = pat.parts[1].pattern
  const [b, r] = innerPat.kind === 'bind' ? [innerPat.name, innerPat.pattern] : [undefined, innerPat]
  const bs2 = partial_match(mod, r, types.partial_eltype(val), path)
  if (bs2 === null || bs2 === undefined) return bs2
  if (bs2.size > 0) return undefined
  return b === undefined ? bs : _assoc(bs, b, [val, path])
}

function partial_match(mod: Interpreter, pat: Pattern, val: types.Type, path: Path = []): MatchResult {
  switch (pat.kind) {
    case 'hole':
      return new Map()

    case 'literal': // TODO use assoc
      if (types.isValue(val) && types.isValue(pat.value)) {
        return isEqual(pat.value, val) ? new Map() : null
      } else {
        return undefined // could reject more cases here
      }

    case 'bind':
      const bs = partial_match(mod, pat.pattern, val, path)
      if (bs === null || bs === undefined) return bs
      return _assoc(bs, pat.name, [val, path])

    case 'trait':
      const r = trivial_isa(mod, val, pat.trait)
      return r === true ? new Map() : r === false ? null : undefined

    case 'or':
      if (val.kind === 'recursive') return partial_match(mod, pat, types.unroll(val), path)
      if (val.kind === 'union') return partial_match_union(mod, pat, val, path)
      for (const p of pat.patterns) {
        const m = partial_match(mod, p, val, path)
        if (m === null) continue
        if (m === undefined) return undefined
        return m
      }
      return null

    case 'constructor': {
      const patType = patternType(pat)
      const result = mod.get(types.tag('common.constructorPattern'), [types.list(...types.parts(patType))])
      if (!result || !types.isValue(result)) return undefined
      return partial_match(mod, pattern(types.part(result, 1)), val, path)
    }

    case 'pack':
      if (types.isAtom(val) || val.kind === 'pack') {
        return partial_match_pack(mod, pat, val, path)
      } else if (val.kind === 'vpack') {
        return partial_match_vpack(mod, pat, val, path)
      } else if (val.kind === 'union') {
        return partial_match_union(mod, pat, val, path)
      } else if (val.kind === 'recursive') {
        return partial_match(mod, pat, types.unroll(val), path)
      } else {
        throw new Error('unimplemented')
      }

    case 'repeat':
    case 'and':
      return undefined

    default:
      const _: never = pat
      throw new Error('unreachable')
  }
}

// TODO assumes the value is unchanged by the match
function trivial_isa(int: Interpreter, val: types.Type, T: types.Type): boolean | undefined {
  const r = int.get(types.tag('common.matchTrait'), [types.list(T, val)])
  if (r === undefined) return undefined
  const tag = types.tagOf(types.part(r, 1))
  if (isEqual(tag, types.tag('common.Some'))) return true
  if (isEqual(tag, types.tag('common.Nil'))) return false
  return undefined
}

// Filtered methods

function MatchMethods(defs: Definitions, interp: Interpreter) {
  return new EagerCache<[types.Tag, types.Type], [Method, Match | undefined][]>(([f, Ts]) => {
    const result: [Method, Match | undefined][] = []
    const methods = defs.methods(f)
    for (const meth of methods.slice().reverse()) {
      const m = partial_match(interp, meth.sig.pattern, Ts)
      if (m === null) continue
      result.push([meth, m])
      if (m !== undefined) break
    }
    return result
  })
}

type MatchMethods = ReturnType<typeof MatchMethods>

// Generate dispatchers

function dispatch_arms(T: types.Type): types.Type[] {
  if (T.kind === 'union') return T.options
  if (T.kind === 'recursive') return dispatch_arms(types.unroll(T))
  if (T.kind === 'pack') {
    let result: types.Type[][] = [[]]
    for (const part of T.parts.map(dispatch_arms))
      result = result.flatMap(prefix => part.map(x => [...prefix, x]))
    return result.map(parts => types.pack(...parts))
  }
  return [T]
}

function indexer(code: MIR, T: types.Type, arg: number, path: Path): number {
  if (path.length === 0) return arg
  const [p, ...rest] = path
  if (typeof p !== 'number') {
    const ps: number[] = []
    for (let i = p.start; i <= p.end; i++)
      ps.push(code.push(code.stmt(xpart(arg, types.Type(BigInt(i))), { type: types.part(T, i) })))
    const L = types.list(...ps.map(v => code.type(v) as types.Type))
    arg = code.push(code.stmt(xlist(...ps), { type: L }))
  } else {
    T = types.part(T, p)
    arg = code.push(code.stmt(xpart(arg, types.Type(BigInt(p))), { type: T }))
  }
  return indexer(code, T, arg, rest)
}

function icall(inf: Inference, code: MIR, sig: Sig, f: IRValue, ...args: (IRValue | number)[]): number {
  if (!(f instanceof Method))
    args = [code.push(code.stmt(xlist(...args),
      { type: types.list(...args.map(a => types.asType(code.type(a)))) }))]
  const ex = xcall(f, ...args)
  const T = inferexpr(inf, sig, code, ex)
  return code.push(code.stmt(ex, { type: T }))
}

function dispatcher(inf: Inference, func: types.Tag, Ts: types.Type): [MIR, ir.Anno<types.Type>] {
  const code = MIR(Def(func.path, undefined, true))
  const args = code.argument(Ts)
  let ret: ir.Anno<types.Type> = ir.unreachable
  let arms = dispatch_arms(Ts)
  const call = (f: IRValue, ...as: (IRValue | number)[]) => icall(inf, code, [func, Ts], f, ...as)
  for (const [meth, m] of inf.meths.get([func, Ts])) {
    if (m === undefined) {
      const pat = patternType(meth.sig.pattern)
      arms = arms.filter(T =>
        issubset(types.list(types.nil), some(infercall(inf, [func, Ts], types.tag('common.match'), types.list(T, pat)))))
      let m = call(types.tag('common.match'), args, pat)
      if (code.type(m) === ir.unreachable) { code.block().unreachable(); return [code, ret] }
      m = call(part_method, m, types.Type(1n))
      // TODO use call?
      const cond = code.push(code.stmt(xcall(isnil_method, m), { type: partial_isnil(types.asType(code.type(m))) }))
      code.branch(code.blockCount + 2, [], { when: cond })
      code.branch(code.blockCount + 1)
      code.newBlock()
      m = call(notnil_method, m)
      const as: number[] = []
      for (const arg of meth.sig.args)
        as.push(call(part_method, call(types.tag('common.getkey'), m, types.tag(arg)), types.Type(1n)))
      let result = call(meth, ...as)
      if (meth.sig.swap.size === 0 && code.type(result) !== ir.unreachable)
        result = code.push(code.stmt(xlist(result), { type: types.list(types.asType(code.type(result))) }))
      code.return(result)
      ret = maybe_union(ret, ir.asAnno(types.asType, code.type(result)))
      code.newBlock()
    } else { // certain to match
      const as = meth.sig.args.map(x => indexer(code, Ts, args, some(m.get(x))[1]))
      let result = call(meth, ...as)
      if (code.type(result) === ir.unreachable) {
        code.block().unreachable()
        return [code, ret]
      }
      if (meth.sig.swap.size === 0)
        result = code.push(code.stmt(xlist(result), { type: types.list(types.asType(code.type(result))) }))
      code.return(result)
      ret = maybe_union(ret, ir.asAnno(types.asType, code.type(result)))
      return [code, ret]
    }
    if (arms.length === 0) {
      code.block().unreachable()
      return [code, ret]
    }
  }
  if (isEqual(func, types.tag('common.abort')) && types.issubset(Ts, types.list(types.String())))
    throw new Error("Compiler fault: couldn't guarantee abort method matches")
  if (options().jspanic)
    call(types.tag('common.abort'), string(code, `No matching method: ${func}: ${types.repr(Ts)}`))
  code.block().unreachable()
  return [code, ret]
}
